### ISSUE TITLE
Nested TypedDict not assignable to Json

### ISSUE THREAD

**Issue by chris-leach** [microsoft/pyright#9254]
I'm trying to pass a nested structure of `TypedDict`s into a function that accepts a JSON value type. It should be evident from the following code that `Parent` is a subtype of `Json`, and Pyright does not return an error for the non-annotated value, but does for the annotated one.

```python
from typing import TypedDict

Json = None | bool | int | float | str | list["Json"] | dict[str, "Json"]

class Child(TypedDict):
    a: str
    b: int

class Parent(TypedDict):
    c: str
    child: Child

def use_json(input: Json) -> None: ...

non_annotated = {"c": "x", "child": {"a": "y", "b": 1}}
use_json(non_annotated)  # no error

annotated: Parent = {"c": "x", "child": {"a": "y", "b": 1}}
use_json(annotated)  # error!
```
Error:
```
.../test_json.py
  .../test_json.py:23:10 - error: Argument of type "Parent" cannot be assigned to parameter "input" of type "Json" in function "use_json"
    Type "Parent" is not assignable to type "Json"
      "Parent" is not assignable to "None"
      "Parent" is not assignable to "bool"
      "Parent" is not assignable to "int"
      "Parent" is not assignable to "float"
      "Parent" is not assignable to "str"
      "Parent" is not assignable to "list[Json]"
      "Parent" is not assignable to "dict[str, Json]" (reportArgumentType)
1 error, 0 warnings, 0 informations
```

**VS Code extension or command-line**
Reproducible with:
Pylance v2024.10.1
Pyright CLI v1.1.385

**Comment by erictraut** [comment]
Pyright is working correctly here, so I don't consider this a bug.

By default, the type of `non_annotated` is inferred as `dict[str, Any]`, which is assignable to `Json`. If you enable `strictDictionaryInference` in your pyright configuration (or enable "strict" type checking), then the type of `non_annotated` will be inferred as `dict[str, str | dict[str, str | int]]`, and it is not assignable to `Json`. For more details, refer to [this documentation](https://microsoft.github.io/pyright/#/type-inference?id=dictionary-expressions).

The type of `annotated` is declared to be `Parent`, and the expression assigned to it is a valid `Parent` type, so its type remains `Parent` after assignment (i.e. there is no opportunity to narrow it further on assignment). The type `Parent` is a TypedDict, and it is not assignable to `Json` according to the rules spelled out in the typing spec. In fact, a `TypedDict` is never assignable to a `dict` of any sort because `dict` types are mutable, which allows the typing guarantees of a `TypedDict` to be violated through mutation. For more details, refer to [this section](https://typing.readthedocs.io/en/latest/spec/typeddict.html) of the typing spec.

I don't have enough context to provide guidance on how to fix your code, but one possible solution is to declare the type of `annotated` as `Json`.

```python
annotated: Json = {"c": "x", "child": {"a": "y", "b": 1}}
use_json(annotated)
```

**Comment by chris-leach** [comment]
Thanks very much for your response. The mutability point is a good one, and I don't need that for my use case.

I tried changing `list` and `dict` to `Sequence` and `Mapping`, and decorating `Parent` with `@final`. 

```python
from collections.abc import Mapping, Sequence
from typing import TypedDict, final

Json = None | bool | int | float | str | Sequence["Json"] | Mapping[str, "Json"]

# ...

@final
class Parent(TypedDict):
# ...
```

I assumed this would be sound, given the only limitation on assignability to `Mapping` mentioned by the spec is due to subclasses, but the issue persists.

I realise this is a duplicate of [#3526](https://github.com/microsoft/pyright/issues/3526) though - a shame, but I'll have to resort to `# pyright: ignore` for now.

Thanks again.
